core/embed/rust/src/ui/display.rs

```diff
@@ -105,6 +151,341 @@ pub fn rect_fill_rounded1(r: Rect, fg_color: Color, bg_color: Color) {
     }
 }
 
+#[derive(Copy, Clone, PartialEq, Eq)]
+pub struct TextOverlay<'a> {
+    colortable: [Color; 16],
+    area: Rect,
+    text: &'a str,
+    font: Font,
+}
+
+impl<'a> TextOverlay<'a> {
+    pub fn new(bg_color: Color, fg_color: Color, text: &'a str, font: Font) -> Self {
+        let area = Rect::zero();
+        Self {
+            colortable: get_color_table(fg_color, bg_color),
+            area,
+            text,
+            font,
+        }
+    }
+
+    pub fn place(&mut self, baseline: Point) {
+        let text_width = self.font.text_width(self.text);
+        let text_height = self.font.text_height();
+
+        let text_area_start = baseline + Offset::new(-(text_width / 2), -text_height);
+        let text_area_end = baseline + Offset::new(text_width / 2, 0);
+        let area = Rect::new(text_area_start, text_area_end);
+
+        self.area = area;
+    }
+
+    pub fn get_pixel(&self, underlying: Option<Color>, p: Point) -> Option<Color> {
```

Comment:
it seems this should always return a `Color` instead of an Option. if the pixel is out of bounds, the underlying color is returned.

also I don't understand the usecase for not specifying an underlying color?

Reply:
changed: https://github.com/trezor/trezor-firmware/commit/ce97b6452904d229e17dc2e672562804e24a71f2

I used the unspecified underlying in the hold to confirm animation - where i didn't want to specify it, as in the caller the color was used only as info whether to negate the color (if the text color differed from underlying). Changed it - removed the colortable from overlay struct and this function gets both underlying and FG color for the text from the caller. This way it can support more usecases (i.e. deciding what color the text should be based on luminance of the underlying - which should be done in the caller function).